codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -104,16 +172,21 @@ public T getAndRemove(T name, T defaultValue) {
     }
 
     @Override
-    @SuppressWarnings("unchecked")
     public List<T> getAll(T name) {
-        Object value = map.get(name);
-        if (isList(value)) {
-            return unmodifiableList((List<T>) value);
-        }
-        if (value == null) {
-            return emptyList();
+        checkNotNull(name, "name");
+
+        LinkedList<T> values = new LinkedList<T>();
```

Comment:
@buchgr - I can't find the other comment on this, but I think my hesitation before was for this case we don't know how many elements there are, and we must insert in the front every time. For an `ArrayList` this would involve a `list.add(0, obj)` which would have to do an array shift each time.  It is not clear to me that this would be a net gain so can we handle in a followup PR if necessary?